test/cmp_testlib.c

```diff
@@ -39,6 +39,7 @@ X509 *load_pem_cert(const char *file)
         (void)TEST_ptr(cert = PEM_read_bio_X509(bio, NULL, NULL, NULL));
```

Comment:
x509_set0_libctx() should only be used as a last resort for resolving..

The better way to do this is 

```
TEST_ptr(cert = X509_new_with_libctx(libctx, propq);
..
PEM_read_bio_X509(bio, &cert, NULL, NULL)
..
```
If you can think of other places where you might of done this it would be good if they were changed..


If you have a look at d2i_X509() you will notice that it calls x509v3_cache_extensions() if the cert is passed in..
This makes it a bit more of a immutable operation.. This was also the reason why I was a bit wary of some of the 
x509v3_cache_extensions() calls that you had added in a few places.

So the rules I used for the CMS changes were..
Use X509_new_with_libctx() everywhere instead of X509_new().
If a ASN1 object got loaded that contains certs.. resolve them straight afterwards..

Reply:
Thanks for your explanation.
I've changed this as you mentioned, and I did not find any further such occurrences.

I still think the calls of  `x509v3_cache_extensions()` I added at some places are needed there
because otherwise I had situations where I got reported misleading errors.

The CMP+CRMF code does not (directly) use `X509_new()`.

Reply:
See below :)

Reply:
Meanwhile I understand better what you mean with doing any needed`x509v3_cache_extensions()` as early as possible.
I've just removed one of those because I found it's done elsewhere.
The remaining two are used in cases where certs are input by a CMP_CTX_set1 function or by the CMP request/response transfer callback which in general is not under the control of the CMP implementation, neither.

Reply:
So the question then becomes - what bit of code is currently loading the offending cert - and make the point where it is loaded where the resolve is done..

Reply:
> So the question then becomes - what bit of code is currently loading the offending cert - and make the point where it is loaded where the resolve is done..

The `x509v3_cache_extensions()` call in `cmp_ctx()` is right in `OSSL_CMP_CTX_set1_##FIELD` where a cert is being imported by the crypto lib user. In the tests the cert has been loaded using the `PEM_read_bio_X509()` call in `cmp_testlib.c` which we have meanwhile handled. So this should be fine now.

The other remaining `x509v3_cache_extensions()` call is in `cmp_vfy()` where the cert has been determined by a non-trivial process called via `ossl_cmp_msg_check_update()` (and then `OSSL_CMP_validate_msg()`) which typically selects it from the `extraCerts` field in a received CMP message. In the test these are loaded via `load_pkimsg()` in `cmp_testlib.c`. This indirectly calls `d2i_OSSL_CMP_MSG_bio()` where you wrote that this invokes  `x509v3_cache_extensions()` for any certs included.

So I'd say the two remaining explicit `x509v3_cache_extensions()` call in the lib are good as precautions in case a lib user does not properly augment the certs (s)he provides with any needed libctx etc.